Use current time() instead of Request Time for the expiration TTL

MatthiasKuehneEllerhold opened the pull request:
The calculation of the expiration timestamp is problematic if you have long-running requests (like user uploads). For long running requests the `$_SERVER['REQUEST_TIME']` is far less than the current timestamp. For example uploading 500 MiB can take up to 4 hours in some of our use cases. We're using `setExpirationSeconds()` as a way to enforce a server-side inactivity timeout. Lets assume we're setting the expiration to 2 hours (`$container->setExpirationSeconds(7200);`) it will set the expiration timestamp to 2 hours from the START of the request (`$_SERVER['REQUEST_TIME']`) and not 2 hours from NOW (`time()`).

Let me show this on an real-life example: 
* We call `$container->setExpirationSeconds(7200)` on every request.
* At timestamp 0 we're starting an upload that takes 4 hours (until 14400).
* We constantly (every 100 seconds) ping our server during the upload to avoid getting the inactivity logout.
  * This resets the expiration timestamp to each pings REQUEST_TIME + 2 hours.
  * Browsing the website during this time is possible without an inactivity logout.
* Timestamp 14300 (100 seconds before the end of the data upload): The expiration timestamp is set via the ping to 21500 (14300 + 7200).
* Timestamp 14400: the data upload is done.
  * It finds an intact session (because REQUEST_TIME is used as comparison and 0 < 21500).
  * It sets the new expiration timestamp to 7200 (REQUEST_TIME + 7200) (down from 21500).
* Timestamp 14401: The next requests sees the expiration timestamp of 7200 (which is 2 hours in the past) and resets all data.

In order to fix this the container needs to use the current time() instead of the REQUEST_TIME for setting the new expiration TTL.

I hope I could make this (rather specific) problem clear.

MatthiasKuehneEllerhold committed:
Use current time() instead of Request Time for the expiration TTL

Ocramius commented:
@MatthiasKuehneEllerhold the problem is clear, but a test (even using `sleep()`, without having to mock the heck out of the internal functions) is required for patches to be considered.

MatthiasKuehneEllerhold committed:
Unit-Tests

MatthiasKuehneEllerhold committed:
Add unit-test

MatthiasKuehneEllerhold commented:
@Ocramius I've fixed the current unit-tests and added a new one. The problem with this kind of tests is that during the execution the second can switch between the calls. So we'd need to make them robust enough against that but they shouldn't take too long either.
My test case *should* fail on the old implementation and work on the new one.

MatthiasKuehneEllerhold commented:
Travis fails on PHP 7.1 on this unit-test: `ZendTest\Session\SaveHandler\MongoDBTest::testReadDestroysExpiredSession`. I dont think this is related to my changes, especially since all other jobs (even the other 7.1 ones) work flawlessly.

MatthiasKuehneEllerhold commented:
@Ocramius Any feedback?

Ocramius commented:
Didn't get to check it at all, sorry: busy with other OSS projects atm

On 22 Aug 2017 8:13 AM, "Matthias Kühne" <notifications@github.com> wrote:

> @Ocramius <https://github.com/ocramius> Any feedback?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendframework/zend-session/pull/85#issuecomment-323929324>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJakAi5bAPZc133Xh5sioXiwIgsoLi2ks5sanGegaJpZM4O6OoE>
> .
>

weierophinney commented on test/ContainerTest.php:
```diff
@@ -217,15 +219,24 @@ public function testMultipleCallsToExpirationSecondsAggregates()
+        $this->assertEquals($currentTimestamp + 3600, $metadata['EXPIRE']);
+    }
+
+    public function testSettingExpirationSecondsUsesCurrentTime()
```
Unfortunately, this test doesn't prove the fix.

To test, I reverted the change you provided (`s/time()/$_SERVER['REQUEST_TIME']/`), and ran just this test; it passed. I tried resetting the sleep and expiration seconds values as well, but they always passed. I also tried putting the assignment _prior_ to the `setExpirationSeconds()` method, and that did not work, either.

While I do not doubt that the fix is correct, we need an accurate way to test it that emulates what happens in a normal server-side web request so that we can ensure the fix is correct and will not break with future updates.

MatthiasKuehneEllerhold committed:
Fix Unit-Test so that it correctly tests for the new behavior and fails on the old

MatthiasKuehneEllerhold commented:
Sorry for missing this! You were correct, the test proved nothing because the REQUEST_TIME never changed. I've fixed it and the REQUEST_TIME will now be temporarily overwritten to simulate a second request coming it.

weierophinney commented on test/ContainerTest.php:
```diff
@@ -217,15 +219,40 @@ public function testMultipleCallsToExpirationSecondsAggregates()
+
+        // Simulate a second request, backup the request time and overwrite it with current time()
+        $requestTimeBackup      = $_SERVER['REQUEST_TIME'];
+        $requestTimeFloatBackup = $_SERVER['REQUEST_TIME_FLOAT'] ?? null;
```
This library still supports PHP 5.6, so we'll need to do an `isset() ? ... : null` here. I can do that on merge, though.

weierophinney commented on test/ContainerTest.php:
```diff
@@ -217,15 +219,40 @@ public function testMultipleCallsToExpirationSecondsAggregates()
+        } catch (\Throwable $e) {
+            // Restore the original values
+            $_SERVER['REQUEST_TIME']       = $requestTimeBackup;
+            $_SERVER['REQUEST_TIME_FLOAT'] = $requestTimeFloatBackup;
```
I don't think we need to do this; PHPUnit restores superglobals between test executions by default, IIRC.

I'll try removing it locally when I merge.

weierophinney commented:
I've verified the fix locally at this time by restoring the original code and running the new tests! Merging momentarily!

MatthiasKuehneEllerhold commented:
Thanks!
